### PR TITLE
Stabilize extern_crate_self

### LINKED ISSUE #56409: Tracking issue for feature `extern_crate_self`
With this feature `extern crate NAME as RENAME;` now accepts `NAME`=`self` and interprets it as referring to the local crate.

As with other `extern crate` items, `RENAME` in this case gets into extern prelude and therefore can be used from absolute paths and imports on 2018 edition.

```rust
extern crate self as serde; // Adds local crate to extern prelude as `serde`

mod foo { ... }

// Anywhere in this crate
use serde::foo; // OK
::serde::foo; // OK
```

The feature's primary purpose is to resolve https://github.com/rust-lang/rust/issues/54647 and support migration of proc macro crates to 2018 edition.

Implementation PR - https://github.com/rust-lang/rust/pull/55275.

### PR DESCRIPTION
**Pull request by mehcode**
Fixes #56409

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -689,6 +686,8 @@ declare_features! (
     (accepted, repr_packed, "1.33.0", Some(33158), None),
     // Allows calling `const unsafe fn` inside `unsafe` blocks in `const fn` functions.
     (accepted, min_const_unsafe_fn, "1.33.0", Some(55607), None),
+    // `extern crate self as foo;` puts local crate root into extern prelude under name `foo`.
+    (accepted, extern_crate_self, "1.33.0", Some(56409), None),
```
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    (accepted, extern_crate_self, "1.34.0", Some(56409), None),
```
This won't make it to 1.33.

**Review by Centril — CHANGES_REQUESTED** [review]
Looks good; the version needs to be adjusted tho.

**Comment by Centril** [comment]
# Stabilization report

r? @Centril 

**I propose that we stabilize `#[feature(extern_crate_self)]` with 1.34 as the earliest stable version.**

@rfcbot merge

+ Proposed in: https://github.com/rust-lang/rust/pull/55275 and accepted by fcp-merge https://github.com/rust-lang/rust/pull/55275#issuecomment-435574377.
+ Implementation PR: https://github.com/rust-lang/rust/pull/55275 (reviewed by: @eddyb)
+ Days since in nightly: 39 (~5.57 weeks)
+ Tracking issue: https://github.com/rust-lang/rust/issues/56409
+ [Feature gate test being removed](https://github.com/rust-lang/rust/blob/789a15a4221acbbbd2ec1352a0a5ce9f497a4425/src/test/ui/feature-gates/feature-gate-extern_crate_self.rs)
+ Documentation issue in reference: https://github.com/rust-lang-nursery/reference/issues/501

## Motivation

To resolve https://github.com/rust-lang/rust/issues/54647. Also see https://github.com/rust-lang/rust/issues/56409#issuecomment-451776791.

As noted in https://github.com/rust-lang/rust/pull/55275#issue-224835313, this solution doesn't introduce any new syntax and has minimal maintenance cost.

## What is stabilized

Suppose that we have the following item in the current crate:

```rust
fn the_answer() -> usize { 42 }
```

It now becomes *semantically* possible ([test](https://github.com/rust-lang/rust/blob/f79eeaa02d6c4fa5f21f69b30aed5e7ccbd21ad7/src/test/ui/imports/extern-crate-self-pass.rs)) (it *already* was *syntactically* possible) to write `extern crate self as ALIAS` where `ALIAS` is an identifier. The semantics of `extern crate self as ALIAS` is `ALIAS` is added to the extern prelude and will refer to the current crate. Since `ALIAS` is now in the extern prelude you may now write `::ALIAS::the_answer` and `ALIAS::the_answer` under the rules due to https://github.com/rust-lang/rust/pull/54658.

Since `ALIAS` is any identifier, you may now for example write: `extern crate self as serde;` and then refer to `the_answer` by `serde::the_answer`.

You may also write:

+ ```rust
  macro_rules! alias_self {
      ($alias:ident) => { extern crate self as $alias; }
  }
  
  alias_self!(the_alias);
  
  fn main() {
      assert_eq!(the_alias::the_answer(), 42);
  }
  ```

+ ```rust
  macro_rules! extern_something {
      ($alias:ident) => { extern crate self as the_alias; }
  }
   
  extern_something!(self);
   
  fn main() {
      assert_eq!(the_alias::hello(), 42);
  }
  ```

  (@mehcode: please extend the set of tests to include both as `// run-pass` tests..)

## What is not stabilized:

+ `extern crate self;` is semantically rejected ([test](https://github.com/rust-lang/rust/blob/f79eeaa02d6c4fa5f21f69b30aed5e7ccbd21ad7/src/test/ui/imports/extern-crate-self-fail.rs)) since it would add `self` to the extern prelude thus coming into conflict with `self::`.

  However, it is already syntactically valid and will remain so.
  For example:

  ```rust
  macro_rules! accept_item { ($x:item) => {} }

  accept_item! {
      extern crate self;
  }
  ```

  (@mehcode: please extend the set of tests to include this as a `// compile-pass` test..)

**Review by Centril — APPROVED** [review]

File: src/test/ui/imports/extern-crate-self-macro-item.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+macro_rules! accept_item { ($x:item) => {} }
```
Comment (Centril) [inline src/test/ui/imports/extern-crate-self-macro-item.rs]: This should be a compile-pass test, the default for UI is compile-fail.

Also add a comment above...

File: src/test/run-pass/imports/extern-crate-self-macro-1.rs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+fn the_answer() -> usize { 42 }
```
Comment (Centril) [inline src/test/run-pass/imports/extern-crate-self-macro-1.rs]: Add a comment above explaining the purpose of the test.

File: src/test/run-pass/imports/extern-crate-self-macro-2.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+fn the_answer() -> usize { 42 }
```
Comment (Centril) [inline src/test/run-pass/imports/extern-crate-self-macro-2.rs]: Add a comment above explaining the purpose of the test.

File: src/test/run-pass/imports/extern-crate-self-macro-1.rs (hunk 3 above)
Comment (Centril) [inline src/test/run-pass/imports/extern-crate-self-macro-1.rs]: Move this file to `ui/imports/` and explicitly write `// run-pass` on the first line.

If you can put all the test files related to `extern crate self` into `imports/extern-crate-self` that would be nice.

File: src/test/run-pass/imports/extern-crate-self-macro-2.rs (hunk 4 above)
Comment (Centril) [inline src/test/run-pass/imports/extern-crate-self-macro-2.rs]: Also move this file to ui...

**Review by Centril — CHANGES_REQUESTED** [review]

**Comment by mehcode** [comment]
Thanks for walking me through this @Centril.

File: src/test/ui/imports/extern-crate-self/extern-crate-self-macro-alias.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// run-pass
+
+// Test that a macro can correctly expand the alias
+// in an `extern crate self as ALIAS` statement
```
Comment (Centril) [inline src/test/ui/imports/extern-crate-self/extern-crate-self-macro-alias.rs]:
```suggestion
// in an `extern crate self as ALIAS` item.
```

File: src/test/ui/imports/extern-crate-self/extern-crate-self-macro-item.rs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// compile-pass
+
+// Test that `extern crate self;` is accepted
+// syntactically as an item for use in a macro
```
Comment (Centril) [inline src/test/ui/imports/extern-crate-self/extern-crate-self-macro-item.rs]:
```suggestion
// syntactically as an item for use in a macro.
```

File: src/test/ui/imports/extern-crate-self/extern-crate-self-macro-self.rs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+// run-pass
+
+// Test that a macro can correctly expand `self` in
+// an `extern crate self as ALIAS` statement
```
Comment (Centril) [inline src/test/ui/imports/extern-crate-self/extern-crate-self-macro-self.rs]:
```suggestion
// an `extern crate self as ALIAS` item.
```

**Review by Centril — CHANGES_REQUESTED** [review]
Some minor style nits; it looks great otherwise.

**Review by Centril — APPROVED** [review]

**Comment by mehcode** [comment]
@Centril What's the next step here?

**Comment by Centril** [comment]
@mehcode Waiting ~two days; after that I'll r+ :)

**Comment by mehcode** [comment]
Thanks for the clarification. Read somewhere FCP was 7 days so I was confused.

**Comment by Centril** [comment]
Hehe ;) It's 10 days.

**Comment by scottmcm** [comment]
@mehcode you're not the only one: https://github.com/rust-lang/rfcs/pull/1981

**Pull request merged**